### PR TITLE
Prototype API for log redaction

### PR DESCRIPTION
**Pull request by varungandhi-src**
Implements what I describe in my [comment earlier](https://github.com/sourcegraph/sourcegraph/issues/33473#issuecomment-1141647021).

The PR is written for commit-by-commit review.

We don't need to merge this if we think that this is the wrong direction to go in, but I figured it'd be easier to understand the proposed changes with actual code rather than without.

(I'm not attached to the specific names of the different types, cases etc; feel free to suggest alternate names.)

## Test plan

Added tests for non-trivial new functions.

### CODE REVIEW

File: lib/log/privacy/text.go [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+
+// Text is a helper type for tracking strings which may potentially
+// be logged and/or contain user-private or org-private information.
+//
+// It acts as an "immutable" container; instead of allowing ad-hoc
+// modification, methods return new Text values.
+type Text struct {
+	data    string
+	privacy Privacy
+}
```
Comment (bobheadxi) [inline lib/log/privacy/text.go:11]: Do we want to implement JSON marshalling and/or `.String()` here?

File: lib/log/fields.go [hunk 2]
```diff
@@ -49,6 +45,44 @@ var (
...
+	panic("Impossible to call!")
+}
+
+// String is a function that should not be implemented. It's presence would increase
+// the risk of accidentally log user-private data, making it visible to a site-admin.
+//
+// Use Texts instead.
+func Strings(key string, nope UseFuncTextsInstead) Field {
+	panic("Impossible to call!")
+}
```
[the remark this answers is not in the document]
Comment (bobheadxi) [reply lib/log/fields.go:84]: I think we can have:

```go
func String(string, privacy.Text) Field
func Strings(string, []privacy.Text) Field

// Shorthand, in case you don't have a privacy.Text available
func Text(string, string, privacy.Privacy) Field
func Texts(string, []string, privacy.Privacy) Field
```

**Review by bobheadxi — COMMENTED** [review]
Thanks for putting this up! 😁 some comments on the API

**Comment by bobheadxi** [comment]
Also https://github.com/sourcegraph/sourcegraph/pull/36005 is almost done, let's make sure that gets merged first since I don't want to block Gitstart on that PR any longer 😆

File: lib/log/fields.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (varungandhi-src) [reply lib/log/fields.go:84]: I think there is a conceptual point here which I have not made super clear.

I see `Text` as type for encouraging code to be "privacy-correct", not just as a helper type for redaction. For that to work, it would make sense to have struct fields etc. also use `Text` where applicable. Not having the shorthands makes using `Text` and plumbing it around the path of least resistance, nudging people to use more `Text` when they're going to log the value. 

This is why I deliberately skipped the shorthands.

---

That said, if you feel strongly that we ought to have the shorthands, I can make that change.

If we do go down that route, I think naming the functions the other way around would be more consistent `Text` function would log `Text` values, and the `String` function would log `string` values (with an extra privacy setting).

File: lib/log/privacy/text.go [hunk 3]
```diff
@@ -0,0 +1,56 @@
+package privacy
```
[the remark this answers is not in the document]
Comment (varungandhi-src) [reply lib/log/privacy/text.go:1]: That works, I wasn't entirely sure of where to put it.

File: lib/log/fields.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (bobheadxi) [reply lib/log/fields.go:84]: Thanks for the clarification on skipping shorthands! While best practices are good, I do think it is important to have an alias for simple string fields needs, at least to help adoption. Reversing it sounds fine though:

```
func Text(string, privacy.Text) Field
func Texts(string, []privacy.Text) Field

// Shorthand, in case you don't have a privacy.Text available
func String(string, string, privacy.Privacy) Field
func Strings(string, []string, privacy.Privacy) Field
```

File: lib/log/privacy/text.go (hunk 1 above)
Comment (bobheadxi) [reply lib/log/privacy/text.go:11]: We might want to document the lack of `String()` here

**Comment by keegancsmith** [comment]
From reading the updates to call sites, this API feels a bit verbose. I see our enum is only two fields, can I propose an alternative implementation:

- remove log.String
- Add log.PublicString
- Add log.PrivateString

I would even consider just making it `log.Public` and `log.Private`. My concern is logging code is everywhere, so succinct names are very valuable here to avoid noise when reading code. Given logging is everywhere, developers will very quickly get up to speed knowing what log.Public/etc are and it will read nicely.

**Comment by bobheadxi** [comment]
With @keegancsmith 's proposal, I think instead of:

> remove log.String

We can keep it as discussed in https://github.com/sourcegraph/sourcegraph/pull/36300#discussion_r885827046 to accept text instead, e.g.

```go
func String(string, privacy.Text) Field
func Strings(string, []privacy.Text) Field
```

I think a way to directly accept `privacy.Text` is nice so that structs, etc can perhaps use it directly elsewhere

**Comment by varungandhi-src** [comment]
> I see our enum is only two fields, can I propose an alternative implementation:

The third case `Unknown` is a deliberate choice for migration: it gives a clear signal on what code needs to be audited to be privacy-correct (instead of being privacy-oblivious). If we only use two cases, replacing existing uses of `log.String` with `Public` or `Private` is problematic because:
- If we switch to `Private` everywhere, we risk losing out on useful public information when investigating incidents.
- If we switch to `Public` everywhere, then auditing code incrementally becomes harder because it is not possible to easily distinguish between "this is deliberately marked Public because this information is public" vs "this was marked Public for historical reasons but needs a further look"

File: lib/log/privacy/text.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (varungandhi-src) [reply lib/log/privacy/text.go]: Moved the package to `lib/privacy`.

File: lib/log/privacy/text.go (hunk 1 above)
Comment (varungandhi-src) [reply lib/log/privacy/text.go]: Documented lack of `String()` and added JSON (un)marshalling with tests.

File: lib/log/fields.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (varungandhi-src) [reply lib/log/fields.go]: Added `String(s)?` shorthands to avoid the explicit call to `NewText`.

**Comment by keegancsmith** [comment]
I'm a little confused on how we will be using this, it seems like the point of `privacy.Text` is we use it instead of strings all over our app? Is that correct?

Can you update some part of this PR so that the updates you did to callsites read like you would expect them to read once this API is everywhere? Right now it is hard to evaluate the proposed APIs, since your call site updates are all just privacy.Unknown. Personally I find ensuring our call sites are still easy to read is important, based on the current callsite updates this feels clunky.

Additionally, is there some context we can read to understand the user expectations here. It would be great to know what we regard as private? How will we debug stuff for example if every repo name is just `<redacted>`? Do we expect admins to tune privacy, since this doesn't do post-hoc reduction, it just never logs the values.

**Comment by bobheadxi** [comment]
> How will we debug stuff for example if every repo name is just <redacted>?

Might be expensive, but I wonder if we can render a hash of the value? Then the debugging use case would still be possible by tracing a particular hash through the logs

File: cmd/gitserver/server/cleanup.go [hunk 4]
```diff
@@ -264,9 +267,9 @@ func (s *Server) cleanupRepos(gitServerAddrs []string) {
 		// name is the relative path to ReposDir, but without the .git suffix.
 		repo := s.name(dir)
 		subCleanupLogger := cleanupLogger.With(
-			log.String("repo", string(repo)),
+			log.String("repo", string(repo), privacy.Unknown),
```
Comment (bobheadxi) [inline cmd/gitserver/server/cleanup.go:270]: As @keegancsmith mentioned, it might be good to have a few examples where we use the API as intended. The [original issue outlines repo names as the main use case](https://github.com/sourcegraph/sourcegraph/issues/33473), so perhaps we can start by change `RepoName` to a `privacy.Text`?

Comment (varungandhi-src) [reply cmd/gitserver/server/cleanup.go:270]: > so perhaps we can start by changing `RepoName` to a `privacy.Text` instead of `string`? Such a change and the impact it might have would probably help inform the design of this API

OK, I can try to do this, but I'm only going to touch a subset of files first since this type is used _everywhere_ throughout the codebase.

**Comment by varungandhi-src** [comment]
> It would be great to know what we regard as private?

This is part of the doc comments.

```
        // Private represents a piece of data that is specific to a user or an
	// organization, which means that it shouldn't be leaked outside
	// (such as to a site-admin via logs).
	Private Privacy = iota
	// Unknown represents a piece of data that is unclear in its status
	// (whether it is Public or Private). It is intended to aid the
	// transition where we redact private information from logs.
	//
	// Avoid introducing new uses of Unknown if possible.
	Unknown
	// Public represents a piece of data that has instance-wide visibility.
	Public
```

At some point in the repo lifecycle, we become aware whether the repo has instance-wide visibility or not (I see there are already `Public bool` fields in some places). If it does, mark the `RepoName` as `Public`. If it is specific to a user or org, mark it `Private`. Until that point in time, mark it as `Unknown`.

There is already some code right now which uses a `Public bool` field, this is similar to that.

Now, I get that the privacy of a repo can change over time, so strictly speaking, there is a TOCTOU problem here, but that aspect seems unavoidable (e.g. even if we literally did a permission check immediately before logging, ignoring perf issues, we could still end up logging something which was made private between it being logged and being seen by a site-admin).

> Might be expensive, but I wonder if we can render a hash of the value instead of <redacted>?

Are you concerned about the cost of hashing itself or something else?

One thing we could do is that if the string is long: only hash a fixed-length prefix and add a length marker. [Hashing a few bytes should be cheaper than an L3 miss](https://twitter.com/pkhuong/status/1531238526750232577?s=20&t=d9_GSxLpNA429p7SG95nUg). Moreover, I'm not 100% sure, but I think since we're mainly trying to do obfuscation rather than trying to prevent a site-admin from "attacking" the instance, it should be OK to use a fast non-cryptographic like FNV-1a?

I suspect allocating a new string and formatting it is likely to be more expensive than computing the hash itself.

**Comment by bobheadxi** [comment]
> Are you concerned about the cost of hashing itself or something else?
>
> One thing we could do is that if the string is long: only hash a fixed-length prefix and add a length marker. [Hashing a few bytes should be cheaper than an L3 miss](https://twitter.com/pkhuong/status/1531238526750232577?s=20&t=d9_GSxLpNA429p7SG95nUg). Moreover, I'm not 100% sure, but I think since we're mainly trying to do obfuscation rather than trying to prevent a site-admin from "attacking" the instance, it should be OK to use a fast non-cryptographic like FNV-1a?
>
> I suspect allocating a new string and formatting it is likely to be more expensive than computing the hash itself.

Was pondering both size and computation, which seem to be potentially non-issues - we can definitely do a fixed-length prefix (don't really need to be that concerned about collision, since "good enough" is probably fine, we don't need strict uniqueness guarantees), and also agree that the goal is obfuscation rather than security guarantees - so a non-cryptographic hash is acceptable as well. If those points are okay, I think we should go ahead and do this approach rather than log `<redacted>` everywhere

**Comment by varungandhi-src** [comment]
https://github.com/sourcegraph/sourcegraph/pull/36300/commits/85650a49fc0953163b23454e165d64cf893f1f8c includes some changes to logging sites after changing the implementation of `RepoName`.

**Comment by varungandhi-src** [comment]
https://github.com/sourcegraph/sourcegraph/pull/36300/commits/702bf6738bdbd47478f5f01cd0638a56a8db1d53 includes printing the hash.

File: lib/log/fields.go [hunk 5]
```diff
@@ -54,6 +51,62 @@ var (
...
+func fnv1a(s string, maxBytes int) uint32 {
+	// See https://en.wikipedia.org/wiki/Fowler%E2%80%93Noll%E2%80%93Vo_hash_function#FNV_hash_parameters
+	hash := uint64(0x811c9dc5)
+	for i := 0; i < maxBytes; i++ {
+		hash = (hash ^ uint64(s[i])) * 0x01000193
+	}
+	return uint32(hash)
+}
+
+func redact(s string) string {
```
Comment (bobheadxi) [inline lib/log/fields.go:67]:
```suggestion

// redact renders a hash of string to aid in debugging, rather than render '<redacted>' everywhere.
// It does not have any uniqueness or security guarantees.
func redact(s string) string {
```

File: internal/api/api.go [hunk 6]
```diff
@@ -34,7 +37,9 @@ var deletedRegex = lazyregexp.New("DELETED-[0-9]+\\.[0-9]+-")
 // change its name in the database, this function extracts the original repo
 // name.
 func UndeletedRepoName(name RepoName) RepoName {
-	return RepoName(deletedRegex.ReplaceAllString(string(name), ""))
+	return RepoName{name.Text.MapData(func(s string) string {
+		return deletedRegex.ReplaceAllString(s, "")
+	})}
```
Comment (bobheadxi) [inline internal/api/api.go:42]: Nifty!

File: internal/api/api.go [hunk 7]
```diff
@@ -60,7 +65,7 @@ type RepoCommit struct {
 
 func (rc RepoCommit) LogFields() []log.Field {
 	return []log.Field{
-		log.String("repo", string(rc.Repo)),
+		log.String("repo", rc.Repo.Text.GetDataUnchecked()),
```
Comment (bobheadxi) [inline internal/api/api.go:68]: Should this be unchecked? This helper gets fields for logging

File: lib/log/fields.go (hunk 5 above)
Comment (bobheadxi) [reply lib/log/fields.go:67]: Also we might want to move these helpers into `redact.go` in this package, to keep `fields.go` easy to read for those seeking an API reference.

**Comment by bobheadxi** [comment]
Overall:

1. `log.Text` API looks okay to me
2. For `log.String`, having to write out `pivacy.Private` is unfortunately quite verbose, and I share Keegan's concerns about readability. Off the top of my head I can't think of much of an alternative though - maybe we can find a shorter name for package `privacy`? maybe `vis` for `visibility`? That might obfuscate the purpose a bit though
3. `api.RepoName` change seems okay to me. Having to do `repo.Text` feels a bit strange, though I suppose you can't add functions on an aliased type 😞 Does this marshal/unmarshal correctly from e.g. the database still?

Once concerns are addressed, we should make sure relevant stakeholders in `api.RepoName` get the chance to review as well (or maybe just all of dev-backend)

File: internal/api/api.go (hunk 7 above)
Comment (varungandhi-src) [reply internal/api/api.go:68]: Yes this should be unchecked, because this is indiscriminately logging the repo name instead of checking the privacy level (this PR is meant as a no-op in terms of semantics). This `log` is actually:
```
	"github.com/opentracing/opentracing-go/log"
```
Not `lib/log`. Otherwise, I would've used `log.Text` here.

**Comment by varungandhi-src** [comment]
> For log.String, having to write out pivacy.Private is unfortunately quite verbose, and I share Keegan's concerns about readability.

If anyone else has suggestions on how to reduce verbosity, I'm all ears. For example, Keegan suggested per-Privacy methods. I can add those if you feel like that'd help to some extent? It would save a few characters for the `privacy.` prefix.

> api.RepoName change seems okay to me. Having to log repo.Text instead of just repo feels a bit strange, though I suppose you can't add functions on an aliased type 😞

There's another option here. If we have:

```
// in package api
type RepoName privacy.Text
// in package log
func Text[T ~struct{ data string; privacy Privacy }](key string, value T) log.Field {...}
```

Then the function should accept both `privacy.Text` and `api.RepoName` values. However, that makes accessing the different `Text` methods on `RepoName` more cumbersome because you'd need `privacy.Text(repo).Method()` instead of `repo.Text.Method()`. I actually tried this approach for a bit locally but the casts felt annoying, so I felt the tradeoff wasn't worth it.

> Maybe instead of embedding Text we can call it Value instead? repo.Value reads more naturally to me.

Sure, we can change that. 👍 

> Also, does this marshal/unmarshal correctly from e.g. the database still?

If we're OK with this direction broadly, I propose that we break off the `api.RepoName` changes into a separate PR. I've only made those changes here as a mock-up. Even without that change, this PR is quite invasive (it changes the type of `log.String`), making it easy to run into merge conflicts + code that doesn't compile because of newly added logging calls.

The `api.RepoName` change is also likely to cause a bunch of merge conflicts given the number of places it is used in.

Combining both invasive changes into one PR seems like a recipe for an unmergeable PR.

We can take care of the marshaling for `api.RepoName` in that PR.

File: lib/log/fields.go (hunk 5 above)
Comment (varungandhi-src) [reply lib/log/fields.go]: Created separate `redact.go` file.

**Comment by varungandhi-src** [comment]
Actually, we could get rid of the `.Text`/`.Value` when logging `api.RepoName` values if we created a separate interface and required that as a generic constraint for `func Text`.

```
// in package api
func (r RepoName) AsText() Text { return r.Value }

// in package log
// AsText is a conversion interface for types that are equivalent to privacy.Text
type AsText interface {
  func asText() privacy.Text
}
func Text[T AsText](key string, value T) log.Field
```

**Comment by bobheadxi** [comment]
> Actually, we could get rid of the .Text/.Value when logging api.RepoName values if we created a separate interface and required that as a generic constraint for func Text.

I'm not terribly well-versed in Go generics yet - could Text itself implement this interface so that the currently proposed usage still works? (`log.Text(privacy.Text{})`)

**Comment by bobheadxi** [comment]
> If we're OK with this direction broadly, I propose that we break off the api.RepoName changes into a separate PR. I've only made those changes here as a mock-up. Even without that change, this PR is quite invasive (it changes the type of log.String), making it easy to run into merge conflicts + code that doesn't compile because of newly added logging calls.

**Comment by keegancsmith** [comment]
I need to absorb this a bit more, I have my concerns around making a core datatype more clunky to use. Is there context to the need here (specific customers, compliance, etc), just so when thinking about how this solves the problem one can consider alternatives. In particular as an admin do I want my logs to just say redacted, or hashed? What happens when something is wrong with a repo I use, then we want to find log messages about it? If the use case is for sharing with us for debugging, it feels like there should be another solution.

If repo is the thing that is private, and everything else is public. This is quite a high cost to pay everytime we want to log a string, given most strings we log are not repo names.

**Comment by varungandhi-src** [comment]
> I need to absorb this a bit more, I have my concerns around making a core datatype more clunky to use.

Sure, I'm not in a rush to merge this. It'd be nice to merge this some time next week, to avoid creating more and more conflicts, but we're not a hard deadline.

That said, I'd urge you to consider the situation from the perspective that we may be using `string`s in way too many places than we ought to be using them in (that was certainly my impression when I started working in the codebase and seeing conversions to/from strings all over the place, without any validation). Earlier, I'm guessing this was done for two reasons:
- Getting things working was more important than doing everything "correctly".
- The lack of generics meant that we couldn't have logging APIs that were both nice and avoided heap allocations.

> In particular as an admin do I want my logs to just say redacted, or hashed?

Hashes allow correlating things across log lines, at the cost of some extra verbosity in logs. That seems like the right tradeoff because you're probably looking at the logs because you want to debug something. Are there some other axes you were considering?

> What happens when something is wrong with a repo I use, then we want to find log messages about it?

I'm not sure who "I" is here... is "I" a non-site admin engineer or is "I" a site-admin?

> If the use case is for sharing with us for debugging, it feels like there should be another solution.

It doesn't have to be shared with us; the hashes should be helpful when site-admins are debugging by themselves too.

> If repo is the thing that is private, and everything else is public

Some other examples of things I've seen in logs which may need to be redacted in some situations (not an expert here, details TBD based on what Security folks say):
- Paths and file names
- Branch and tag names (commit SHAs are OK)
- CLI arguments for certain invocations
- Code (snippets)
- Search query strings

> This is quite a high cost to pay everytime we want to log a string, given most strings we log are not repo names

If we use dedicated types for different things instead of using `string`s everywhere, this problem mostly goes away because you can add a `AsText` method to your type which always uses `privacy.Public`.

So the cost would largely be a one-time cost when you modify the type definition; logging values of that type would be as convenient as before. The `RepoName` mock-up is a good example of this.

---

> Is there context to the need here (specific customers, compliance, etc), just so when thinking about how this solves the problem one can consider alternatives.

@deflncha can probably provide more context on this. The reason I brought it up was because our current approach to logging is (if I'm being honest) a little careless, and this kind of change seems easier to do sooner rather than later. We've already had situations internally where someone accidentally pasted private repo names into incident channels in Slack.

I can definitely think of specific existing customers ([such as this customer](https://github.com/sourcegraph/customer/issues/922)) where it would be a problem if site-admins had access to private repo names in logs, even if the repos are using code names.

Addressing this (possibly on a tight deadline) _after_ a customer has flagged it as a concern doesn't seem appealing.

**Comment by keegancsmith** [comment]
Ok, so we have a use case where the logs themselves need to be redacted before being written to `os.Stderr`. Previous experiences I have had with logging at big companies we normally did redaction for sharing / scrubbing for compliance reasons. This would happen afterwards as needed. The situation where an admin of sourcegraph at a customer can't know the name of a repo seems rare? I supposed managed instances are different, but in those cases I would still be surprised since we need to be able to inspect these systems (and have some sort of controls in place to prevent abuse). My instincts align with this, redaction is something you do afterwards, not with how the logs end up at rest. Additionally my instincts say whats the point of logging something if you can't look at it. Why do we log?

FYI if we want to more often post process logs, I think an alternative is instead to mark the key somehow that it is private. With structured logging it will then be easy to write a redacter.

As to the point around using strings all over the place, in the go ecosystem (and many others) that is normal, and I wouldn't call it careless. They are strings. Personally I think you need a really good business need to introduce friction as fundemental as "strings are careless". This might be one such case for some pieces of data. An example of a random thing that breaks when making a change like this is marshalling `[]api.RepoName` for an RPC (or many other things).

**Comment by varungandhi-src** [comment]
I feel like there are several separate questions/concerns here; I'm going to try to break them apart and respond to them individually.

### The API design / verbosity

This was a concern earlier -- are you still concerned about this, or is it OK now?

### Is it useful to log redacted strings at all

If we statically knew that values of a certain type were always going to be redacted, there is certainly a case to be made that we shouldn't log them.

If we only know that information dynamically (as is the case with repo names, code snippets etc), there are two options: (1) log only when the information is not redacted OR (2) always log. This PR selects option (2) because with the suggestion of including the hash, I thought having at least some information for debugging would be helpful to correlate different log lines.

### Should redaction happen at the time of logging or after-the-fact

My understanding that the right "boundary" for redaction was at the point where the log leaves "our system" and goes somewhere else. E.g. if we're sending logs to Datadog, we'd need to redact before sending the logs to Datadog. I wrote this PR based on that understanding/assumption.

That said, I'm not 100% sure about this; I will double-check with security.

---

I would like to keep this PR review largely focused on the specific bits about the redaction API and approach, and not go into a lot more detail on what should be done with specific bits of data. As I mentioned earlier, there are several pieces of information, such as code snippets and so on, which are potentially applicable here.

That said, I think it makes sense to at least discuss `api.RepoName` to some extent, since that was the example we started with and is part of the mock-up here.

### Should api.RepoName values be redacted or not

> The situation where an admin of sourcegraph at a customer can't know the name of a repo seems rare? I supposed managed instances are different, but in those cases I would still be surprised since we need to be able to inspect these systems (and have some sort of controls in place to prevent abuse)

I have some half-baked answers to this, but I'm going to check in with security folks first to make sure I'm on the right track. @Strum355 mentioned "some legal technicalities or something about private information ending up in non-sourcegraph controlled places (datadog etc)" -- it'd be useful to get clarification here.

FWIW, right now, there seems to be some inconsistency on how we hide this information from site-admins. In the code intel uploads GUI, we do [filter out private repos](https://sourcegraph.com/github.com/sourcegraph/sourcegraph/-/blob/internal/codeintel/stores/dbstore/uploads.go?L433%3A2=&utm_source=VSCode-2.2.3). However, we're clearly logging the same information in multiple places.

### Should api.RepoName's definition be adjusted to use this new type or not

> An example of a random thing that breaks when making a change like this is marshalling []api.RepoName for an RPC (or many other things).

I'm not very familiar with RPC stuff, so I don't know what the best fix for this is. But it seems like a broader problem (and not specific to this PR) if we cannot evolve types used in the codebase.

**Comment by jhchabran** [comment]
_Tl;DR great exploration, love the discussion, but strongly feel we need more clarity on the use case_

Thanks for taking the time to write such detailed explanations! I'm trying to go back to the root of the initiative, and I'm having a hard time pin pointing the root discussion. I saw @varungandhi-src's [question](https://sourcegraph.slack.com/archives/C1JH2BEHZ/p1654256407158389) to the sec team about this, which I hope will bring clarity on the exact use case this prototype is targeting. 

My main concern is the price we'd be paying by introducing friction as such a low level. My understanding is that depending on the instance type, the privacy expectations are varying and my first instinct would be to tie the redaction to the instance type rather than within the code. 

As we are focused on providing on-prem and managed instances, I see two cases there: either the customer is the admin of the instance and therefore already have access to everything which is fine because that's their own stuff, or it's a managed instance, where only explicitly authorised teammates have access to logs and PI.  

Please correct me if I'm wrong, but in both cases, the issues lies within exporting the logs, whether it's to our teams for debugging or when the data is sent to a third party. 

When it comes to PI concerns, the most important thing is usually to be certain that we're not leaking anything. And while the current approach here is solid, it relies on the the correctness of the code to ensure that it's not leaking. And as @varungandhi-src pointed out, there is always the TOCTOU problem. 

If we were using post-processing, which have the huge advantage of not messing with strings, we could not only rely on markers to redact variables if we want to, but even better we could also leverage the fact that the processor living where the PI is available, could simply forbid strings by looking at the various bits of data that should be redacted. No more TOCTOU. This approach is much more mechanical and therefore by working purely on the exported content would redact even things we would have missed. The cost would be paid in plumbing, which at first look is much cheaper.

**Comment by varungandhi-src** [comment]
> I'm having a hard time pin pointing the root discussion

There isn't really a "root discussion" as much as this [Slack discussion](https://sourcegraph.slack.com/archives/C07KZF47K/p1648486800725899). One of the mistakes I made here is that I incorrectly assumed (based on a couple of brief convos) that the point about repo names being redacted was already a point that was "agreed-upon-by-everyone" and the only thing lacking was someone rolling up their sleeves and doing the refactoring. Clearly, that's assumption was incorrect 😅; I should've started a more detailed discussion before jumping in with a PR.

> My main concern is the price we'd be paying by introducing friction as such a low level.

IMO, this concern should largely be alleviated if the relevant strings have their own types and are passed around consistently as that type (with methods for functionality that is repeatedly used, instead of direct string manipulation on an as-needed basis), with conversion at boundaries (network, database etc.). As the `RepoName` example shows, once you've changed types, the logging-specific code is no more verbose than before.

In the meantime, during transition, yes, there will be more verbosity (at least with the current APIs). But that can also be reduced if added some of the APIs that @keegancsmith suggested (`log.Public` etc.) which add only a small amount of friction over `log.String`.

But of course, I'm biased, because I wrote this PR the way it is.

> My understanding is that depending on the instance type, the privacy expectations are varying and my first instinct would be to tie the redaction to the instance type rather than within the code.

I'm afraid I don't understand this point. Let's say we're talking about managed instances specifically. Say that a company has some public OSS GitHub repos and some (org-)private GitHub repos; and all of these are added to their managed instance.

From a logging POV, it seems OK if information about the public GitHub repos is logged, since anyone can access them anyways? The same point doesn't seem to apply to the private repos. So it does seem like there is a distinction at the repo level.

Is your point that if we need to redact information about private repos in such a situation anyways, we should redact information about public repos too? (If so, that seems like it would make debugging even harder but maybe we collectively think that's an OK price to pay for simpler code.)

> either the customer is the admin of the instance and therefore already have access to everything which is fine because that's their own stuff

For this specific point, I'm not so sure about it. The customer isn't a single person; different people may have different access levels within the company. There is an underlying assumption here that an admin has access to the source code of every repo at the company. Keegan's point was that this seemed reasonable; based on my experience, I don't think think that it is necessarily true.

As one data point, at least some of the code-intel code (such as the one showing upload info in the GUI) seems to account for this scenario. 

This is one of the things we need input on; that's why I've mentioned in the questions I've posted in the security channel.

**Comment by coury-clark** [comment]
> There is an underlying assumption here that a site admin has access to all repos in the instance. Keegan's point was that this seemed reasonable; based on my experience, I don't think that it is necessarily true.

I think this is true at the code search level. It may not be a fair representation of the company policy though.

Will just say that from the perspective of an on-prem feature, this would probably make logs pretty worthless for us. Though to be fair, they are already pretty worthless to us, so I'm actually not sure there is any practical difference. 

What _would_ be a difference is in an environment such as k8s-dogfood though. This is our primary testing / debugging ground, and where we would definitely want to have unredacted logs. I'd strongly advocate for a way we can configure this redaction per deployment.

**Comment by bobheadxi** [comment]
> I'd strongly advocate for a way we can configure this redaction per deployment.

We could introduce an env variable configuration to toggle how the field gets encoded

**Comment by jhchabran** [comment]
> From a logging POV, it seems OK if information about the public GitHub repos is logged, since anyone can access them anyways? The same point doesn't seem to apply to the private repos. So it does seem like there is a distinction at the repo level.

This is the blurry part that I'd like to see clarified. Does the customer wants to redact repo names so they can give access to the logs to individuals that should not be able to see the repos? Is so that we can't see those when debugging stuff? 

> Is your point that if we need to redact information about private repos in such a situation anyways, we should redact information about public repos too? (If so, that seems like it would make debugging even harder but maybe we collectively think that's an OK price to pay for simpler code.)

No, I don't think we should redact more than we should, that would be making debugging painful, we both agree on that 😊 

> For this specific point, I'm not so sure about it. The customer isn't a single person; different people may have different access levels within the company. There is an underlying assumption here that a site admin has access to all repos in the instance. Keegan's point was that this seemed reasonable; based on my experience, I don't think that it is necessarily true.

> This is one of the things we need input on; that's why I've mentioned in the questions I've posted in the security channel.

Yes, that's mainly my point. My concern there is that we would be moving toward introducing some redaction without establishing clearly when and what to redact. 

> Will just say that from the perspective of an on-prem feature, this would probably make logs pretty worthless for us. Though to be fair, they are already pretty worthless to us, so I'm actually not sure there is any practical difference.

This, we're working on our side in improving the logs themselves, so I want to be careful in going into the other direction. 

> What would be a difference is in an environment such as k8s-dogfood though. This is our primary testing / debugging ground, and where we would definitely want to have unredacted logs. I'd strongly advocate for a way we can configure this redaction per deployment.

100% agree. Otherwise, that would be a huge hit. 

@varungandhi-src I don't want to make it feel like I'm blocking the initiative, which taken as it is, is merely about changing a single type into another in the codebase and is not impacting that much how we fundamentally deal with strings. Instead, I just want to underline the fact that we need clarity on the underlying use case, and how it could impact which solution we deem to be best to solve it :)

**Comment by varungandhi-src** [comment]
I'm closing this PR for now, since we don't have enough clarity on business requirements.

Based on the response by @vrto in [Slack](https://sourcegraph.slack.com/archives/C1JH2BEHZ/p1654604328785629?thread_ts=1654592552.640429&cid=C1JH2BEHZ), it sounds like we'll have more information here a bit later, perhaps by the end of this month. 

We can re-open this PR or build on top of it later if it roughly matches what we need.

**Comment by bobheadxi** [comment]
> Based on the response by @vrto in [Slack](https://sourcegraph.slack.com/archives/C1JH2BEHZ/p1654604328785629?thread_ts=1654592552.640429&cid=C1JH2BEHZ), it sounds like we'll have more information here a bit later, perhaps by the end of this month.

Related update: https://github.com/sourcegraph/sourcegraph/issues/33473#issuecomment-1148929021

**Comment by vrto** [comment]
Catching up on this work, I'm generally quite happy with what I've seen, and my hunch is that we'll need something fairly similar to this.

Redaction will be a fundamental part of our Observability strategy, but _what specifically_ will be redacted should be owned by the teams designing the data. Think of it as security - we can have guidelines, recommendations, and linter checks, but ultimately we must make it a part of our culture and have the people with the most domain knowledge be accountable for redactions within their focus vertical.

I strongly agree that the redaction must happen at the latest right before exiting the system boundary (unless we figure out some kind of _clever (TM)_ way of, I dunno, wrapping stderr with some redaction-aware logic).
